### PR TITLE
add a callback to modify generated entity

### PR DESCRIPTION
**Pull request by berlingoqc**
I want to be able to modify the generated entity based on the api return by the backend.

Here an exemple, i don't know if you have other idea if we want to change some field.

```
const entityMapper = async (entity: Entity, api: any) => {

  const { lifecyclePhase } = api.productVersionMetadata;

  if (entity.spec && lifecyclePhase) {
    entity.spec.lifecycle = lifecyclePhase;
  }

  return entity;
};

export const catalogGlooPlatformPortalBackendProvider = createBackendModule({
  pluginId: 'catalog',
  moduleId: 'gloo-platform-portal-backend-provider',
  register(env) {
    env.registerInit({
      deps: {
        catalog: catalogProcessingExtensionPoint,
        logger: coreServices.logger,
        config: coreServices.rootConfig,
        scheduler: coreServices.scheduler,
      },
      async init({ catalog, logger, config, scheduler }) {
        catalog.addEntityProvider(
          new GlooPlatformPortalProvider(logger, config, scheduler, entityMapper),
        );
      },
    });
  },
});
```

### CODE REVIEW

File: plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts [hunk 1]
```diff
@@ -94,16 +98,24 @@ export class GlooPlatformPortalProvider implements EntityProvider {
...
 
+
   //
   // 1. Init class
   //
   constructor(
     logger: LoggerService,
     config: Config,
     scheduler: SchedulerService,
+  );
```
Comment (Charlesthebird) [inline plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts]: I think we can remove this.

```suggestion
```

File: plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts [hunk 2]
```diff
@@ -308,13 +320,15 @@ export class GlooPlatformPortalProvider implements EntityProvider {
           // For "gloo-gateway"
           //
           this.updatePortalServerType('gloo-gateway');
-          entities.push(await this.getGlooGatewayApiEntity(api));
+          let entity = await this.getGlooGatewayApiEntity(api);
+          entities.push(await this.applyEntityMapping(entity, api));
```
Comment (Charlesthebird) [inline plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts]:
```suggestion
          entities.push(await this.applyEntityTransformation(entity, api));
```

File: plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts [hunk 3]
```diff
@@ -308,13 +320,15 @@ export class GlooPlatformPortalProvider implements EntityProvider {
...
         } else if ('apiProductId' in api) {
           //
           // For "gloo-mesh-gateway"
           //
           this.updatePortalServerType('gloo-mesh-gateway');
-          entities.push(await this.getGlooMeshGatewayApiEntity(api));
+          let entity = await this.getGlooMeshGatewayApiEntity(api);
+          entities.push(await this.applyEntityMapping(entity, api));
```
Comment (Charlesthebird) [inline plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts]:
```suggestion
          entities.push(await this.applyEntityTransformation(entity, api));
```

File: plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts [hunk 4]
```diff
@@ -529,4 +543,12 @@ export class GlooPlatformPortalProvider implements EntityProvider {
 
     return processedAPIs;
   }
+
+  private async applyEntityMapping(entity: Entity, api: ApiVersionExtended | API) {
```
Comment (Charlesthebird) [inline plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts]:
```suggestion
  private async applyEntityTransformation(entity: Entity, api: ApiVersionExtended | API) {
```

File: plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts [hunk 5]
```diff
@@ -529,4 +543,12 @@ export class GlooPlatformPortalProvider implements EntityProvider {
...
+    if (this.entityTransformation) {
```
Comment (Charlesthebird) [inline plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts]: In our UI code, we try to be explicit when checking variable truthiness by converting it to a boolean first, like this (`!!value` is the same as `Boolean(value)`).

```suggestion
    if (!!this.entityTransformation) {
```

**Review by Charlesthebird — COMMENTED** [review]
Thank you for the PR! I was able to test it out and modify the generated entities.

I left some small cleanup/renaming feedback. If you would like to accept those suggestions I can approve.

File: plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts (hunk 1 above)
Comment (berlingoqc) [reply plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts]: done

File: plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts (hunk 2 above)
Comment (berlingoqc) [reply plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts]: done

File: plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts (hunk 4 above)
Comment (berlingoqc) [reply plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts]: done

File: plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts (hunk 5 above)
Comment (berlingoqc) [reply plugins/platform-portal-backstage-plugin-backend/src/provider/GlooPlatformPortalProvider.ts]: done , thanks a lot

**Review by Charlesthebird — APPROVED** [review]
I tested it again after the changes and everything is looking good!

I used this entity transformation to test it in `packages/backend/src/index.ts`:

```ts

import { Entity } from '@backstage/catalog-model';

const entityMapper = async (entity: Entity, api: any) => {
  entity = {
    ...entity,
    metadata: {
      ...entity.metadata,
      title: 'example-' + (entity?.metadata?.title ?? ''),
    },
  };
  return entity;
};
// ...
// And then passed it into the GlooPlatformPortalProvider.
          new GlooPlatformPortalProvider(
            logger,
            config,
            scheduler,
            entityMapper,
          ),
```
Screenshot 2024-10-28 at 11 32 49 AM

**Comment by Charlesthebird** [comment]
@berlingoqc I published a new version of the plugin, so this is in `v0.0.34`, and updated the [installation instructions](https://www.npmjs.com/package/@solo.io/platform-portal-backstage-plugin-backend).

**Pull request merged**